README.md
[![Build Status](https://status.continuousphp.com/git-hub/ec-europa/platform-dev?token=4df2e996-5362-486e-b409-84527de6a65b&branch=develop)](https://continuousphp.com/git-hub/ec-europa/platform-dev)

# NextEuropa

## Requirements

* Composer
* PHP Phar extension
* PhantomJS (in order to run JavaScript during Behat tests)

## Install build system

Before we can build the NextEuropa platform we need to install the build system
itself. This can be done using [composer](https://getcomposer.org/):

```
$ composer install
```

## Customize build properties

Create a new file in the root of the project named `build.properties.local`
using your favourite text editor:

```
$ vim build.properties.local
```

This file will contain configuration which is unique to your development
machine. This is mainly useful for specifying your database credentials and the
username and password of the Drupal admin user so they can be used during the
installation.

Because these settings are personal they should not be shared with the rest of
the team. Make sure you never commit this file!

All options you can use can be found in the `build.properties.dist` file. Just
copy the lines you want to override and change their values. For example:

```
# The location of the Composer binary.
composer.bin = /usr/bin/composer

# The install profile to use.
platform.profile.name = multisite_drupal_standard

# Database settings.
drupal.db.name = my_database
drupal.db.user = root
drupal.db.password = hunter2

# Admin user.
drupal.admin.username = admin
drupal.admin.password = admin

# The base URL to use in Behat tests.
behat.base_url = http://nexteuropa.local
```

## Listing the available build commands

You can get a list of all the available Phing build commands ("targets") with a
short description of each target with the following command:

```
$ ./bin/phing
```

## Building a local development environment

```
$ ./bin/phing build-platform-dev
$ ./bin/phing install-platform
```

## Running Behat tests

The Behat test suite is located in the `tests/` folder. When the development
version is installed (by running `./bin/phing build-platform-dev`) the Behat
configuration files (`behat*.yml`) will be generated automatically using the
base URL that is defined in `build.properties.local`.

If you are not using the development build but one of the other builds
(`build-platform-dist` or `build-multisite-dist`) and you want to run the tests
then you'll need to set up the Behat configuration manually:

```
$ ./bin/phing setup-behat
```

In order to run JavaScript in your Behat tests, you must launch a PhantomJS
instance before. Use the `--debug` parameter to get more information. Please
be sure that the webdriver's port you specify corresponds to the one in your
Behat configuration (`behat.wd_host.url: "http://localhost:8643/wd/hub"`).

```
$ phantomjs --debug=true --webdriver=8643
```

If you prefer to use an actual browser with Selenium instead of PhantomJS, you
need to define the Selenium server URL and browser to use, for instance:

```
behat.wd_host.url = http://localhost:4444/wd/hub
behat.browser.name = chrome
```

The tests can also be run from the root of the repository (or any other folder)
by calling the behat executable directly and specifying the location of the
`behat.yml` configuration file.

Behat tests can be executed from the repository root by running:

```
$ ./bin/behat -c tests/behat.yml
```

With a single Phing task, you can run every tests suites:

```
./bin/phing behat
```

If you want to execute a single test, just provide the path to the test as an
argument. The tests are located in `tests/features/`. For example:

```
$ ./bin/behat -c tests/behat.yml tests/features/content_editing.feature
```

Some tests need to mimic external services that listen on particular ports, e.g.
the central server of the Integration Layer. If you already have services running
on the same ports, they will conflict. You will need to change the ports used in
build.properties.local.

Remember to specify the right configuration file before running the tests.

## Running PHPUnit tests

Custom modules and features can be tested against a running platform installation
by using PHPUnit. When the development version is installed (by running
`./bin/phing build-platform-dev`) the PHPUnit configuration file `phpunit.xml`
will be generated automatically using configuration properties defined in
`build.properties.local`.

If you are not using the development build but one of the other builds
(`build-platform-dist` or `build-multisite-dist`) and you want to run PHPUnit tests
then you'll need to set up the PHPUnit configuration manually by running:

```
$ ./bin/phing setup-phpunit
```

Each custom module or feature can expose unit tests by executing the following steps:

- Add `registry_autoload[] = PSR-4` to `YOUR_MODULE.info`
- Create the following directory: `YOUR_MODULE/src/Tests`
- Add your test classes in the directory above

In order for test classes to be autoloaded they must follow the naming convention below:

- File name must end with `Test.php`
- Class name and file name must be identical
- Class namespace must be set to `namespace Drupal\YOUR_MODULE\Tests;`
- Class must extend `Drupal\nexteuropa\Unit\AbstractUnitTest`

The following is a good example of a valid unit test class:

```php
<?php

/**
 * @file
 * Contains \Drupal\nexteuropa_core\Tests\ExampleTest.
 */

namespace Drupal\nexteuropa_core\Tests;

use Drupal\nexteuropa\Unit\AbstractUnitTest;

/**
 * Class ExampleTest.
 *
 * @package Drupal\nexteuropa_core\Tests
 */
class ExampleTest extends AbstractUnitTest {
  ...
}
```

PHPUnit tests can be executed from the repository root by running:

```
$ ./bin/phpunit -c tests/phpunit.xml
```


## Checking for coding standards violations

When a development build is created by executing the 'build-platform-dev' Phing
target PHP CodeSniffer will be set up and can be run with the following
command:

```
# Scan all files for coding standards violations.
$ ./bin/phpcs

# Scan only a single folder.
$ ./bin/phpcs profiles/common/modules/custom/ecas
```


CHANGELOG.md
# NextEuropa Platform change log

## Version 2.3.0

### Improvements
  * NEPT-4 - Upgrade - chr module to version 7.x-1.8 (2015-Dec-04)
  * NEPT-36 - Performance - Remove css and js injectors from the platform codebase
  * NEPT-61 - DGT connector - Add website identifier in the title of the job sent via poetry
  * NEPT-62 - DGT connector - Make job deadline field mandatory & do not suggest any date
  * NEPT-69 - DGT connector - Add behat scenario to test allowing to resend a rejected job with an increment in version
  * NEPT-71 - DGT connector - Add COUNTER scenarios into behat tests
  * NEPT-93 - Performance/Varnish - Log into watchdog when web frontend caches are cleared
  * NEPT-98 - DGT connector - Allow to add more languages to an ongoing translation request
  * NEPT-112 - Contextual links are not rendered correctly in beans when listed via Views d.o/node/2084823
  * NEPT-124 - Upgrade - SA-CONTRIB-2016-050 - Flag
  * NEPT-130 - Upgrade - Display suite to 7.x-2.14 (2016-Apr-26)
  * NEPT-145 - Tracked changes - Icons shoud always be active when tracked changes exist. (code fix & add test)
  * NEPT-161 - Upgrade - menu_token to beta7 d.o/node/2838033
  * NEPT-163 - EU Login - Replace ECAS labels with new labels to align to EU login
  * NEPT-165 - Ensure PHP 7 compatibility of core platform specific code
  * NEPT-173 - Performance - Do not enable the apachesolr_multisitesearch module by default
  * NEPT-174 - Performance - Make apachesolr optional
  * NEPT-183 - Upgrade - Platform to Drupal 7.51
  * NEPT-206 - Permission 'use media wysiwyg' forced in multisite_wysiwyg feature. Move it from hard to soft config
  * NEPT-220 - Remove module multisite_review from the codebase
  * NEPT-228 - Remove possibility to add "Editorial team member" user role using the bulk update
  * NEPT-257 - Add cce_basic_config readme
  * NEPT-258 - Remove profiles/multisite_drupal_standard/inject_data.php
  * NEPT-287 - Upgrade - UUID module to beta2
  * NEPT-297 - Upgrade - Security risk: 13/25 Upgrade bootstrap to latest version
  * NEPT-357 - Upgrade - i18n_views to version 7.x-3.0-alpha1 + Patches if necessary
  * NEPT-424 - Include the path "/tests" in the CodeSniffer configuration file
  * NEPT-525 - Remove apachesolr, css_injector, js_injector and contact from communities profile
  * NEPT-553 - Nexteuropa_newsroom upgraded
  * NEPT-391 - Remove Flexible Purge module
  * NEPT-1071 - Implement the new header search bars component &  Force the activation of the 'nexteuropa_europa_search' feature

### New features (For more details on each feature, see the README.md file at the root of the feature's folder)
  * NEPT-79 - Performance/Varnish - An administrator can enable the web frontend cache control feature through Feature set
  * NEPT-80 - Performance/Varnish - Automatically clear the web frontend cache of particular content when an action is performed on it
  * NEPT-81 - Performance/Varnish - Eliminate automatic full web frontend cache purges
  * NEPT-182 - DGT connector - As a CEM user I can enable a feature for DGT connector through feature set
  * NEPT-202 - Tracked changes - Create a feature for the functionality provided in ckeditor_lite
  * NEPT-403 - Contact module should be optional, create a feature that can be enabled/disabled
  * NEPT-444 - Add CDN module to the Core stack
  * NEPT-451 - Add Nagios and nexteuropa_nagios module to the Core stack

### Bug fixes
  * NEPT-41 - Users are no longer gaining Editorial Team Member role. Fix issue & add a regression test
  * NEPT-64 - Theme - ec-resp is missing tag $attributes in template block.tpl.php d.o/node/569362
  * NEPT-102 - DGT connector - A page that was already requested for translation can be modified & requested again, even if not all previous translations were received.
  * NEPT-107 - Multilingual - Views & entity translation issue when estonian language is active
  * NEPT-121 - Simplenews - issue with entity cache : Multiple transmissions of newsletter possible. d.o/node/2801239
  * NEPT-123 - Multilingual - Error on tokens for multilingual content
  * NEPT-141 - Tracked Changes - Full HTML + Change tracking doesn't have the right filters enabled
  * NEPT-156 - Theme - ec_resp_link can generate links with an empty class attribute
  * NEPT-185 - Patch chr module to avoid corruption of files during migration. d.o/node/2816399
  * NEPT-211 - Do not generate token for types that are not defined
  * NEPT-212 - Fix issue on parent uri token for taxonomy terms
  * NEPT-221 - WYSIWYG - Description lists (dt) are not correctly rendered because they are not whitelisted in security_allowed_tags
  * NEPT-224 - Fix empty field label for required multiple textfield d.o/node/980144
  * NEPT-255 - multisite_drupal_core - Missing user_default.png and empty_gallery.png files at install
  * NEPT-291 - Nexteuropa_editorial - Addition of ckeditor_lite css file breaks display suite layout
  * NEPT-325 - mpdf version 7.0 was deleted from github
  * NEPT-412 - Security - XSS issue in node form on title fields
  * NEPT-552 - Fix trait not found in nexteuropa_multilingual.behat.inc

## Previous versions
Please refer to archived Delivery notes in the FPFIS wiki regarding previous versions.


install_subsite.sh
#!/bin/bash

#----------------#
#    INCLUDES    #
#----------------#

script_dir=$(readlink -f $(dirname "$O"))

if [ ! -f "${script_dir}/config.sh" ]; then
	echo "No config file found at ${script_dir}/config.sh"
	exit 255
else
source "${script_dir}/config.sh"
fi
if [ ! -f "${script_dir}/functions.sh" ]; then
	echo  "No functions file found at ${script_dir}/functions.sh"
	exit 255
else
source "${script_dir}/functions.sh"
fi

#---------------------#
#     ARGUMENTS GET   #
#---------------------#

usage="Installation of multisite instance\n
Syntax : $(basename $0) [ARGS] SITE-NAME\n
\t-?,-h, --help\t\tPrint this message\n
\t-p, \tDefine the parent installation name\n
\t-i, \tDefine the installation profile to use the parent site\n
\t-s, \tDefine the subfolder for a new site instance\n
\t-c, \tDefine the custom subsite (projet) name\n
\t-v, --verbose\t\tSet the script in verbose mode\n
\t-f, --force\t\tForce the installation without any request to user input\n
\t\t\t\t(Take care, it might delete automatically your database)\n
\t-d, \t\t\tDefine drush options\n
"

# Configuration of the script
while getopts "i:b:t:d:p:c:vrsfhk?-:" option; do
        #Management of the --options
        if [ "$option" = "-" ]; then
                case $OPTARG in
                        help) option=h ;;
						drush_options)=d ;;
                        verbose) option=v ;;
						force) force=f ;;
						install_profile) option=i ;;
						subdirectory) option=s ;;
						subsite_parent_name) option=p ;;
						subsite_custom_subsite) option=c ;;
                        *)
                                echo "[ERROR] Unknown option --$OPTARG"
                                exit 1
                                ;;
                esac
        fi
        case $option in
				d) drush_options=$OPTARG ;;
                v) verbose=1 ;;
				f) force=1 ;;
				i) install_profile=$OPTARG ;;
				s) subdirectory=$OPTARG ;;
				p) subsite_parent_name=$OPTARG ;;
				c) subsite_custom_subsite=$OPTARG ;;
                \?|h)
                        echo -e $usage
                        exit 0
                        ;;
                :)
                        echo "[ERROR] Missing arguments for -$OPTARG"
                        exit 1
                        ;;
                ?)
                        echo "[ERROR] Unknown option -$option"
                        exit 1
                        ;;
        esac
done

#------------------------#
#     ARGUMENTS CHECK    #
#------------------------#

# site name
subsite_name=$BASH_ARGV
[ -z "${subsite_name}" ] && _exit_with_message 230 "You must provide a subsite."

# parent site name
[ -z "${subsite_parent_name}" ] && _exit_with_message 240 "You must provide the name of the parent_name Drupal instance."

# Set up parent directory
if [ -n "${subdirectory}" ]; then 
	parent_dir="${webroot}/${subdirectory}/${subsite_parent_name}"
	base_path="${base_path}/${subdirectory}/${subsite_parent_name}/${subsite_name}"
else
	parent_dir="${webroot}/${subsite_parent_name}"
	base_path="${base_path}/${subsite_parent_name}/${subsite_name}"
fi

# set DB connection
db_name="${subsite_parent_name}_${subsite_name}"
db_url="mysqli://${db_user}:${db_pass}@${db_host}:${db_port}/${db_name}"
__echo "Set DB URL to mysqli://${db_user}:DB_PASS@${db_host}:${db_port}/${db_name}"
_create_database

# Generate the URL pattern for the provided subsite name
subsite_url_pattern=$(printf "$subsite_cluster_url_pattern" "${subsite_name}")
subsite_dir=$(printf "${subsite_cluster_dir}" "${subsite_name}")

# cleanup target directory
if [ -d "${parent_dir}/sites/${subsite_dir}" ] ; then
	__echo "The folder '${parent_dir}/sites/${subsite_dir}' already exist, it will be deleted" 'warning'
	_continue
	#chmod 744 -R "${parent_dir}/sites/${subsite_dir}"
	rm -Rf "${parent_dir}/sites/${subsite_dir}"
	__echo "Removing the folder '${parent_dir}/sites/${subsite_dir}' done" 'status'
fi

# Set up the drush option
if [ "${force}" = 1 ] ; then
	drush_options="${drush_options} -y"
fi
__echo "Set drush options: ${drush_options}"

__echo "chdir to ${parent_dir}/sites/"


cd "${parent_dir}/sites" || _exit_with_message 145 "Unable to chdir into ${parent_dir}/sites"
# /!\ now working in the "sites" subdirectory of the master site directory


#--------------------------#
#   BUILD SUBSITE SOURCES  #
#--------------------------#

# create subsite directory
__echo "creating directory ${subsite_dir}"
mkdir -p ${subsite_dir} || _exit_with_message 140 "Unable to create ${subsite_dir} in $(pwd)"

# generating settings.php 
__echo "generating ${subsite_dir}/settings.php from default.settings.php"
cp "default/default.settings.php" "${subsite_dir}/settings.php" || _exit_with_message 130 "Unable to create ${subsite_dir}/settings.php"
__fix_perms "${subsite_dir}/settings.php"

# setup files directory 
__echo "creating 'files' subdirectory for ${subsite_name}"
local_subsite_files_directory="${subsite_dir}/files"
__fix_perms "${local_subsite_files_directory}"
mkdir -p "${local_subsite_files_directory}" || _exit_with_message 120 "Unable to create ${local_subsite_files_directory} in $(pwd)"
mkdir -p "${local_subsite_files_directory}/private_files" || _exit_with_message 118 "Unable to create ${local_subsite_files_directory}/private_files in $(pwd)"
__fix_perms ${local_subsite_files_directory}
# get default file from parent site
__echo "Filling 'files' subdirectory with default images"
cp -a "default/files/default_images" "${subsite_dir}/files/"

# fill list of subsite on site.php 
if [  "$subsite_update_drupal_sites_list" == "yes" ]; then
	# we must provide Drupal with the "URL pattern to conf directory" mapping
	
	if [ "$subsite_drupal_sites_list" != "sites.php" ]; then
		# we will not provide this information into the standard sites.php
		# therefore, we may want to provide our own sites.php file
		if [ -f "${script_dir}/sites.php" ]; then
			__echo "Overwriting sites.php"
			cp "${script_dir}/sites.php" "sites.php"
		fi
	fi
	sites_php=${subsite_drupal_sites_list:-sites.php}
	__echo "Filling ${sites_php}"
	[ ! -f "${sites_php}" ] && echo '<?php' > "${sites_php}"
	printf "\$sites['%s'] = '%s';\n" "$subsite_url_pattern" "${subsite_dir}" >> "${sites_php}"
fi

__echo "creating ${subsite_name} subsite itself"
cd ..
if [ ! -s "${subsite_name}" ]; then
	ln -s . "${subsite_name}"
fi


# get custom subsite from SVN
if [ -n "${subsite_custom_subsite}" ]; then 
	svn_files=(
		"modules"
		"libraries"
		"themes"
		"files"
	)
	svn_path="${svn_url}/trunk/custom_subsites/${subsite_custom_subsite}"
	__echo "SVN repository path set to ${svn_path}"
	own_source_path="./${subsite_dir}"
	__echo "own_source_path path set to ${own_source_path}"
	_get_svn_sources
fi

#----------------------#
#   INSTALL SUBSITE    #
#----------------------#

# /!\ now working in the master site directory
drush si $install_profile --sites-subdir="${subsite_dir}" --db-url="${db_url}" --account-name="$account_name" --account-pass="$account_pass" --site-name="$subsite_name" --site-mail="$site_mail"

# echo "completing settings.php"
__fix_perms "sites/${subsite_dir}"
__fix_perms "sites/${subsite_dir}/settings.php"
# perl one-liner printing its entire stdin at the second line of the parsed file
#perl -i -ple 'BEGIN{ @l = <STDIN>; } print @l if ($. == 2)' "sites/${subsite_dir}/settings.php" <<EOF
#\$multisite_subsite = '${subsite_name}';
#include(dirname(__FILE__) . '/../settings.common.php');
#EOF

#-------------------#
#   CONFIG SUBSITE  #
#-------------------#

cd "sites/${subsite_dir}"

#solR config
_setsolrconf

#inject data
_inject_data "${parent_dir}/profiles/${install_profile}/inject_data.php"

#solr indexation
__echo "Run solr indexation.."
drush ${drush_options} solr-index

#run cron
__echo "Run drupal cron..."
drush ${drush_options} cron

#remove links from the linkchecker
_setlinkcheckerconf

__echo "Subsite '${subsite_name}' has been created. You can access it by using  'http://$parent_url/$subsite_name'